cpp/ql/src/semmle/code/cpp/controlflow/IRGuards.qll

```diff
@@ -0,0 +1,482 @@
+    override predicate comparesLt(Expr left, Expr right, int k, boolean isLessThan, boolean testIsTrue) {
+      exists(Instruction li, Instruction ri |
+        remove_conversions(li.getAST()) = left and
+        remove_conversions(ri.getAST()) = right and
+        ir.comparesLt(li, ri, k, isLessThan, testIsTrue)
+      )
+    }
+
+    /** Holds if (determined by this guard) `left < right + k` must be `isLessThan` in `block`.
+        If `isLessThan = false` then this implies `left >= right + k`.  */
+    override predicate ensuresLt(Expr left, Expr right, int k, BasicBlock block, boolean isLessThan) {
+      exists(Instruction li, Instruction ri, boolean testIsTrue |
+        remove_conversions(li.getAST()) = left and
+        remove_conversions(ri.getAST()) = right and
+        ir.comparesLt(li, ri, k, isLessThan, testIsTrue) and
+        this.controls(block, testIsTrue)
+      )
+    }
+
+    /** Holds if (determined by this guard) `left == right + k` evaluates to `areEqual` if this expression evaluates to `testIsTrue`. */
+    override predicate comparesEq(Expr left, Expr right, int k, boolean areEqual, boolean testIsTrue) {
+      exists(Instruction li, Instruction ri |
+        remove_conversions(li.getAST()) = left and
+        remove_conversions(ri.getAST()) = right and
+        ir.comparesEq(li, ri, k, areEqual, testIsTrue)
+      )
+    }
+
+    /** Holds if (determined by this guard) `left == right + k` must be `areEqual` in `block`.
+        If `areEqual = false` then this implies `left != right + k`.  */
+    override predicate ensuresEq(Expr left, Expr right, int k, BasicBlock block, boolean areEqual) {
+        exists(Instruction li, Instruction ri, boolean testIsTrue |
+        remove_conversions(li.getAST()) = left and
+        remove_conversions(ri.getAST()) = right and
+        ir.comparesEq(li, ri, k, areEqual, testIsTrue)
+        and this.controls(block, testIsTrue)
+      )
+    }
+
+    /**
+     * Holds if this condition controls `block`, meaning that `block` is only
+     * entered if the value of this condition is `testIsTrue`. This helper
+     * predicate does not necessarily hold for binary logical operations like
+     * `&&` and `||`. See the detailed explanation on predicate `controls`.
+     */
+    private predicate controlsBlock(BasicBlock controlled, boolean testIsTrue) {
+      exists(IRBlock irb |
+        forex(IRGuardCondition inst | inst = ir | inst.controls(irb, testIsTrue)) and
+        irb.getAnInstruction().getAST().(ControlFlowNode).getBasicBlock() = controlled
+      )
+    }
+}
+
+/**
+ * A Boolean condition in the IR that guards one or more basic blocks. This includes
+ * operands of logical operators but not switch statements. Note that `&&` and `||`
+ * don't have an explicit representation in the IR, and therefore will not appear as
+ * IRGuardConditions.
+ */
```

Comment:
I don't think there's a useful place to put those: you'd want them to be unconditional, but they also need to depend on their left operand which is only executed conditionally...

Reply:
I'd put Nop where we put the logical operators in the CFG of AST nodes: before its operands. So the `a && b` in `if (A && B)` would get the IR CFG `Nop -> A -> [ more edges ]`, where the Nop instruction's `getAST()` is `&&`.

Reply:
@dave-bartolomeo @jbj Do we want to make these changes, and if so, do we want to do it as part of this PR?

Reply:
On further thought, I don't think that adding `Nop`s to represent these values will actually improve our analysis of the IR. It will make mapping back to the AST easier, but @dave-bartolomeo had an alternate proposal that involved tagging the `ConditionalBranchInstruction`s with the logical expression that they are associated with.

Reply:
@rdmarsh2 you're right, I don't think the changes we're discussing here need to block this PR.